File: Foundation/lec18/lastocc.cpp
#include<iostream>
#include<vector>
using namespace std;

int lastocc(vector<int>& arr,int i,int data,int loc){
    if(i==arr.size()){
        return loc;
    }
       if(arr[i]==data){
          loc=i;               
       }     
       int recans=lastocc(arr,i+1,data,loc);  
       return recans; 
}

int main(){
    vector<int> arr{2,7,5,8,5,9,10};
    cout<<lastocc(arr,0,5,-1);
    return 0;
}

File: Foundation/lec09/lastocc.java
class lastocc{
    public static int cf(int[] arr,int data){
      int left=0,right=arr.length-1;
      
      while(left<=right){
          int mid=(left+right)/2;
          if(arr[mid]==data){
                  if(mid+1<arr.length && data==arr[mid+1])
                  {
                    left=mid+1;
                  } 
                  else{
                    return mid;
                  }
          }
          else if(data<arr[mid]){
             right=mid-1;            
          }else{
             left=mid+1; 
          }
          }
  
   return -1; 
  }
  
  
  
  
  
    public static void main(String[] args){
  
      int[] arr={1,1,1,1,2,3,5,5};
      System.out.println(cf(arr,2));
  }
  }